nucypher/cli/commands/dkg.py

```diff
@@ -0,0 +1,343 @@
+import random
+import time
+
+import click
+import maya
+from nucypher_core.ferveo import DkgPublicKey
+from web3 import Web3
+
+from nucypher.blockchain.eth.agents import (
+    ContractAgency,
+    CoordinatorAgent,
+    PREApplicationAgent,
+)
+from nucypher.blockchain.eth.registry import InMemoryContractRegistry
+from nucypher.blockchain.eth.signers import Signer
+from nucypher.characters.lawful import Bob, Enrico
+from nucypher.cli.options import option_eth_provider_uri
+from nucypher.crypto.powers import TransactingPower
+from nucypher.policy.conditions.lingo import ConditionLingo
+from nucypher.utilities.emitters import StdoutEmitter
+from nucypher.utilities.logging import GlobalLoggerSettings
+
+GlobalLoggerSettings.start_console_logging()
+emitter = StdoutEmitter(verbosity=2)
+
+
+option_eth_staking_network = click.option(
+    "--eth-staking-network",
+    "eth_staking_network",
+    help="ETH staking network",
+    type=click.Choice(["tapir", "lynx"]),
+    default="lynx",
+)
+
+option_coordinator_provider = click.option(
+    "--coordinator-provider",
+    "coordinator_provider_uri",
+    help="Coordinator network provider URI",
+    type=click.STRING,
+    required=True,
+)
+
+option_coordinator_network = click.option(
+    "--coordinator-network",
+    "coordinator_network",
+    help="Coordinator network",
+    type=click.Choice(["mumbai"]),
+    default="mumbai",
+)
+
+
+def get_agency(
+    coordinator_network, coordinator_provider_uri, eth_staking_network, eth_provider_uri
+):
+    coordinator_agent = ContractAgency.get_agent(
+        agent_class=CoordinatorAgent,
+        registry=InMemoryContractRegistry.from_latest_publication(
+            network=coordinator_network
+        ),
+        provider_uri=coordinator_provider_uri,
+    )  # type: CoordinatorAgent
+
+    staking_network_registry = InMemoryContractRegistry.from_latest_publication(
+        network=eth_staking_network
+    )
+    application_agent = ContractAgency.get_agent(
+        agent_class=PREApplicationAgent,
+        registry=staking_network_registry,
+        provider_uri=eth_provider_uri,
+    )  # type: PREApplicationAgent
+
+    return coordinator_agent, application_agent, staking_network_registry
+
+
+@click.group()
+def dkg():
+    """DKG management commands"""
+
+
+@dkg.command()
+@option_eth_provider_uri(required=True)
+@option_eth_staking_network
+@option_coordinator_provider
+@option_coordinator_network
+@click.option(
+    "--signer",
+    "signer_uri",
+    "-S",
+    help="Signer URI for initiating a new ritual with Coordinator contract",
+    default=None,
+    type=click.STRING,
+)
+@click.option(
+    "--dkg-size",
+    "dkg_size",
+    "-n",
+    help="Number of nodes to participate in ritual",
+    type=click.INT,
+    default=0,
+)
+@click.option(
+    "--num-rituals",
+    "num_rituals",
+    "-r",
+    help="The number of rituals to initiate",
```

Comment:
This option in particular doesn't make much sense for a `nucypher` command. `nucypher` is still a user-facing application (to node operators), and we're exposing here pretty much development-only functionalities

Reply:
The script was fine, no? An alternative is the `hidden=True` kwarg for click options.

Reply:
The script outgrew itself, it evolved into a full click application.  This merely formalizes it's usage.  If your understanding of the `nucypher` CLI is that it is *only* meant for node operators, then why have the `ursula` namespace at all?

Reply:
My point is that the CLI can potentially be used for operations other than node operation, zooming out even more, this codebase is not just about nodes.  It also has a full-blown python API for TACo.

Reply:
> Why indeed!

Correct me if I'm misreading here, but it feels to me like you're saying, without actually saying, that you don't think that the other Characters and utilities warrant having CLI tooling built for them.

If that is your position, let's have a larger (but please, not protracted) discussion about that.  I think it's very obvious that this utility is asking to be made into a CLI-command, given its trajectory.  So if we want to force it in a different direction, it needs to be part of a larger notion of what this interface represents and to whom.

Reply:
I must admit, initially I was also pro on moving the command as done in this PR, but I am reconsidering now, and I think @cygnusv 's argument is a good one.

While it is slightly awkward to have it in `scripts/` because it uses `click`, it was clear that it was dev functionality like relocking dependencies and the other scripts in that folder.

Moving the dkg script to a `nucypher` command means that this happens:
```bash
$ nucypher --help
Usage: nucypher [OPTIONS] COMMAND [ARGS]...

  Top level command for all things nucypher.

Options:
  --version       Echo the CLI version
  --config-path   Echo the configuration root directory path
  --logging-path  Echo the logging root directory path
  --help          Show this message and exit.

Commands:
  dkg     DKG management commands
  ursula  "Ursula the Untrusted" PRE Re-encryption node management commands.
  ```
which is strange because the `dkg` command is just for testing, and can lead to confusion.
  
If the use of `click` is specifically the main onus for moving, originally I had both the creation of rituals,and the encryption/decryption testing of existing rituals as one command and used   `sys.argv[<index>` which got complicated because the parameters and the relationships between them became too complex (see #3169). Hence the use of  `click` to make this simpler - but I probably should have just split the script into two separate scripts anyway: one for ritual creation, one for testing existing rituals. This is evidenced by the work in this PR to split that functionality into two different commands (`init`,`check` respectively), which is a good thing.

One option is to have two separate scripts in the `scripts/` folder, and therefore we may not even need to use `click` anyway since `sys.argv[<index>]` would be sufficient because the complexities of the parameters and their relationships would be reduced. Of course, we could still use `click` or we could have something like:
- `dkg_init <eth-provider> <coordinator-provider> <num_rituals> <dkg_size> <signer> [optional<eth-staking-network>]
- `dkg_check <eth-provider> <coordinator-provider> <ritual_id> [optional<eth-staking-network>]